### PR TITLE
etcdmain: support sdnotify for readiness

### LINKED ISSUE #2724: etcd should notify systemd when it's ready to accept requests
If etcd sent a notify signal to systemd via [sd_notify](http://www.freedesktop.org/software/systemd/man/sd_notify.html), then we could make the service a "notify" type. That way people wouldn't need to write various implementations of wait-for-etcd.service.

### PR DESCRIPTION
**Pull request by xiang90**
Fix #2724 

/cc @crawford @eyakubovich

### CODE REVIEW

**Comment by xiang90** [comment]
@crawford @yichengq Can you please give it a test to see if this matches your expectation? Thanks!

**Comment by crawford** [comment]
This works as intended. The three second pause is going to be problematic though. I assume you have a proper fix on the way.

**Comment by xiang90** [comment]
@crawford I removed the 3s waiting. See comments for details. There is no way to know the status of the inner loop of HTTP server. But it would take normally less than 10ms for the http server routine to be scheduled and starting to serve the first connection.

File: etcdmain/etcd.go [hunk 1]
```diff
@@ -134,6 +135,21 @@ func Main() {
...
 
+	if systemdutil.IsRunningSystemd() {
+		go func() {
+			// At this point, the initialization of etcd is done.
+			// The listeners are listening on the TCP ports and ready
+			// for accepting connections.
+			// The http server is probably ready for serving incoming
+			// connections. If it is not, the connection might be pending
+			// for less than one second.
+			err := daemon.SdNotify("READY=1")
```
Comment (crawford) [inline etcdmain/etcd.go]: does this still need to happen in a goroutine?

Comment (xiang90) [inline etcdmain/etcd.go]: oh. it does not. good catch.

Comment (xiang90) [inline etcdmain/etcd.go]: fixed.

**Comment by crawford** [comment]
LGTM for use on CoreOS. I haven't considered how this behaves on other OSs.

**Comment by xiang90** [comment]
@crawford This should work on the os with system as its init system, right? I am going to merge it. We will have a documentation to talk about the systemd integration and its limitation (ready does not mean the cluster is healthy and this local server is guaranteed to submit request successfully, etc..).

**Pull request merged**